### PR TITLE
feat: mongoose instrumentation

### PR DESCRIPTION
**Pull request by nirsky**

### CODE REVIEW

File: packages/instrumentation-mongoose/src/mongoose.ts [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+    }
+
+    private patchAggregateExec() {
+        const thisInstrumentation = this;
+        thisInstrumentation._logger.debug('mongoose instrumentation: patched mongoose Aggregate exec prototype');
+        return (originalExec: Function) => {
+            return function exec(this: any) {
+                const parentSpan = this[_STORED_PARENT_SPAN];
+                const attributes = {
+                    [DatabaseAttribute.DB_STATEMENT]: thisInstrumentation._config.dbStatementSerializer('aggregate', {
```
Comment (blumamir) [inline packages/instrumentation-mongoose/src/mongoose.ts:103]: what happens if `dbStatementSerializer` returns undefined?
Will we clean this attribute from the span at some point?

File: packages/instrumentation-mongoose/src/mongoose.ts [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+                    return thisInstrumentation._callOriginalFunction(() =>
+                        originalOnModelFunction.apply(this, [
+                            options,
+                            (err: Error, response: any) => {
+                                if (err) {
+                                    setErrorStatus(span, err);
+                                } else {
+                                    safeExecuteInTheMiddle(
+                                        () => thisInstrumentation?._config?.responseHook(span, response),
+                                        () => {},
```
Comment (blumamir) [inline packages/instrumentation-mongoose/src/mongoose.ts:214]: write error to log?

**Review by blumamir — APPROVED** [review]

File: packages/instrumentation-mongoose/LICENSE [hunk 3]
```diff
@@ -0,0 +1,8 @@
+Copyright 2021 @ Aspecto
```
[the remark this answers is not in the document]
Comment (nirsky) [reply packages/instrumentation-mongoose/LICENSE:1]: nope, we checked this specifically with the lawyers

File: packages/instrumentation-mongoose/README.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+| Options        | Type                                   | Description                                                                                     |
+| -------------- | -------------------------------------- | ----------------------------------------------------------------------------------------------- |
+| `suppressInternalInstrumentation` | `boolean` | Mongoose operation use mongodb under the hood. Setting this to true will hide the underlying mongodb spans (if instrumented). |
+| `responseHook` | `MongooseResponseCustomAttributesFunction` | Hook called before response is returned, which allows to add custom attributes to span.      |
+| `dbStatementSerializer` | `DbStatementSerializer` | Mongoose instrumentation will serialize `db.statement` using the specified function.
+
+### Custom `db.statement` Serializer
+
+By default, this instrumentation does not populate the `db.statement` attribute.
```
[the remark this answers is not in the document]
Comment (nirsky) [reply packages/instrumentation-mongoose/README.md:50]: according to specification db.statement is not set by default on mongodb spans 🤷🏼

File: packages/instrumentation-mongoose/README.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    condition?: any;
+    options?: any;
+    updates?: any;
+    document?: any;
+    aggregatePipeline?: any;
+}
+
+type DbStatementSerializer = (operation: string, payload: SerializerPayload) => string;
+```
+Please make sure `dbStatementSerializer` is error proof, as errors are not handled while executing this function.
```
[the remark this answers is not in the document]
Comment (nirsky) [reply packages/instrumentation-mongoose/README.md:65]: same as redis, I followed the pattern but decided to document it.

[the remark this answers is not in the document]
Comment (blumamir) [reply packages/instrumentation-mongoose/README.md:65]: cool :)

File: packages/instrumentation-mongoose/src/mongoose.ts (hunk 1 above)
Comment (nirsky) [reply packages/instrumentation-mongoose/src/mongoose.ts:103]: according to the unit tests it's cleared automartically

File: packages/instrumentation-mongoose/src/mongoose.ts (hunk 2 above)
Comment (nirsky) [reply packages/instrumentation-mongoose/src/mongoose.ts]: done

**Pull request merged**